vms/platformvm/state/state.go

```diff
@@ -924,23 +923,17 @@ func (s *state) SetCurrentSupply(subnetID ids.ID, cs uint64) {
 }
 
 func (s *state) ValidatorSet(subnetID ids.ID, vdrs validators.Set) error {
-	for nodeID, validator := range s.currentStakers.validators[subnetID] {
-		staker := validator.validator
-		if err := vdrs.Add(nodeID, staker.PublicKey, staker.TxID, staker.Weight); err != nil {
-			return err
-		}
+	var err error
+	validatorSet := s.currentStakers.getOrCreateSubnetValidators(subnetID)
 
-		delegatorIterator := NewTreeIterator(validator.delegators)
-		for delegatorIterator.Next() {
-			staker := delegatorIterator.Value()
-			if err := vdrs.AddWeight(nodeID, staker.Weight); err != nil {
-				delegatorIterator.Release()
-				return err
-			}
+	validatorSet.Ascend(func(i *Staker) bool {
+		if err := vdrs.Add(i.NodeID, i.PublicKey, i.TxID, i.Weight); err != nil {
+			return false
 		}
-		delegatorIterator.Release()
-	}
-	return nil
+		return true
+	})
+
+	return err
```

Comment:
@joshua-kim question: are these two implementations equivalent? It seems to me the new version fails to duly account for delegators weight.
If I understand correctly, `getOrCreateSubnetValidators` contains only validators (not delegators) of the specified subnet. The Ascend attribute only walk the tree in-order and execute the vdrs.Add, adding the validators, but does nothing with delegators (which aren't in the collections).

Reply:
A better alternative may be to expose a GetSubnetStakerIterator method in stakers interface, which holds in a tree both validators and delegators for a given subnet.
This may be helpful in other places in our codebase